File: apps/loomx-api/src/routes/dashboards.ts
import { Router, type IRouter } from 'express';
import { DashboardsService } from '../services/dashboards.service';
import { FavoritesService } from '../services/favorites.service';
import { asyncHandler, NotFoundError, ValidationError } from '../middleware/errorHandler';
import { getCurrentUserId } from '../middleware/userContext';
import type { CreateDashboardDTO, UpdateDashboardDTO } from '@loomx/types';

const router: IRouter = Router();
const service = new DashboardsService();
const favoritesService = new FavoritesService();

router.get('/', asyncHandler(async (req, res) => {
  const dashboards = await service.list();

  // Prevent caching - list pages should always be fresh
  res.set({
    'Cache-Control': 'no-cache, no-store, must-revalidate',
    'Pragma': 'no-cache',
    'Expires': '0'
  });

  res.json(dashboards);
}));

router.get('/summary', asyncHandler(async (req, res) => {
  const userId = getCurrentUserId(req);
  const dashboards = await service.list(userId);

  // Prevent caching - list pages should always be fresh
  res.set({
    'Cache-Control': 'no-cache, no-store, must-revalidate',
    'Pragma': 'no-cache',
    'Expires': '0'
  });

  res.json({ count: dashboards.length, recent: dashboards });
}));

router.get('/:id', asyncHandler(async (req, res) => {
  // Disable browser caching - LIVE data only
  res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
  res.setHeader('Pragma', 'no-cache');
  res.setHeader('Expires', '0');

  const dashboard = await service.getById(req.params.id);
  if (!dashboard) throw new NotFoundError('Dashboard');

  // Include is_favorite for the requesting user so the view page can show the correct star state
  const userId = getCurrentUserId(req);
  const isFav = userId !== 'anonymous'
    ? await favoritesService.isFavorite(userId, 'dashboard', req.params.id)
    : false;

  res.json({ ...dashboard, is_favorite: isFav });
}));

router.post('/', asyncHandler(async (req, res) => {
  const data: CreateDashboardDTO = req.body;
  if (!data.name) {
    throw new ValidationError('Name is required');
  }
  const userId = getCurrentUserId(req);
  const dashboard = await service.create(data, userId);
  res.status(201).json(dashboard);
}));

router.put('/:id', asyncHandler(async (req, res) => {
  const data: UpdateDashboardDTO = req.body;
  const dashboard = await service.update(req.params.id, data);
  if (!dashboard) throw new NotFoundError('Dashboard');
  res.json(dashboard);
}));

router.delete('/:id', asyncHandler(async (req, res) => {
  const deleted = await service.delete(req.params.id);
  if (!deleted) throw new NotFoundError('Dashboard');
  res.status(204).send();
}));

/**
 * PUT /api/dashboards/:id/favorite
 * Set or unset favorite status for the current user.
 * Body: { is_favorite: boolean }
 * Respects the desired state rather than blindly toggling (avoids race conditions
 * when the user clicks rapidly or when the client optimistically updates the UI).
 */
router.put('/:id/favorite', asyncHandler(async (req, res) => {
  const userId = getCurrentUserId(req);
  const { id } = req.params;
  const { is_favorite } = req.body;

  const dashboard = await service.getById(id);
  if (!dashboard) throw new NotFoundError('Dashboard');

  if (is_favorite) {
    const favorite = await favoritesService.create(
      { object_type: 'dashboard', object_id: id, object_name: dashboard.name },
      userId
    );
    res.json({ favorited: true, favorite });
  } else {
    await favoritesService.delete(userId, 'dashboard', id);
    res.json({ favorited: false });
  }
}));

export default router;

File: apps/loomx-api/routers/dashboards.py
"""Dashboards router — /api/v1/dashboards."""

from fastapi import APIRouter, Request, Response, HTTPException, Depends
from middleware.auth import require_auth
from models.dashboards import DashboardCreate, DashboardUpdate, DashboardFavoriteBody
import services.dashboards as svc
import services.favorites as fav_svc

router = APIRouter()
NO_CACHE = {
    "Cache-Control": "no-cache, no-store, must-revalidate",
    "Pragma": "no-cache",
    "Expires": "0",
}


@router.get("/dashboards")
def list_dashboards(response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    return svc.list_dashboards()


@router.get("/dashboards/summary")
def dashboards_summary(response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    items = svc.list_dashboards(user)
    return {"count": len(items), "recent": items}


@router.get("/dashboards/{dashboard_id}")
def get_dashboard(dashboard_id: str, response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    dashboard = svc.get_dashboard_by_id(dashboard_id)
    if not dashboard:
        raise HTTPException(status_code=404, detail="Dashboard not found")
    is_fav = fav_svc.is_favorite(user, "dashboard", dashboard_id)
    return {**dashboard, "is_favorite": is_fav}


@router.post("/dashboards", status_code=201)
def create_dashboard(data: DashboardCreate, user: str = Depends(require_auth)):
    return svc.create_dashboard(data.model_dump(exclude_none=True), user)


@router.put("/dashboards/{dashboard_id}")
def update_dashboard(dashboard_id: str, data: DashboardUpdate, user: str = Depends(require_auth)):
    result = svc.update_dashboard(dashboard_id, data.model_dump(exclude_none=True))
    if not result:
        raise HTTPException(status_code=404, detail="Dashboard not found")
    return result


@router.delete("/dashboards/{dashboard_id}", status_code=204)
def delete_dashboard(dashboard_id: str, user: str = Depends(require_auth)):
    deleted = svc.delete_dashboard(dashboard_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Dashboard not found")


@router.put("/dashboards/{dashboard_id}/favorite")
def set_dashboard_favorite(dashboard_id: str, data: DashboardFavoriteBody, user: str = Depends(require_auth)):
    dashboard = svc.get_dashboard_by_id(dashboard_id)
    if not dashboard:
        raise HTTPException(status_code=404, detail="Dashboard not found")

    if data.is_favorite:
        favorite = fav_svc.create_favorite(
            {"object_type": "dashboard", "object_id": dashboard_id, "object_name": dashboard.get("name")},
            user,
        )
        return {"favorited": True, "favorite": favorite}
    else:
        fav_svc.delete_favorite(user, "dashboard", dashboard_id)
        return {"favorited": False}
